cohttp-lwt/src/client.ml

```diff
@@ -12,33 +14,340 @@ module Make (IO : S.IO) (Net : S.Net with module IO = IO) = struct
+      (* give reader a chance to read possible EOF. *)
+      Lwt.pause () >>= Lwt.pause >>= Lwt.pause >>= fun () ->
+      Lwt_mutex.with_lock pipeline.mutex begin fun () ->
+        match pipeline.conn with
+        | Full (_, oc) ->
+          Queue.push {req; body; res_r; res} pipeline.in_flight;
+          Lwt_condition.broadcast pipeline.condition ();
+          Lwt.catch
+            begin fun () -> (* try *)
+                Request.write (fun writer ->
+                    Body.write_body (Request.write_body writer) body
+                  ) req oc
+            end
+            begin fun e -> (* with *)
+              (try Net.close_out oc with _ -> ());
+              Lwt.wakeup_later_exn res_r e;
+              Lwt.return_unit
+            end
+        | Half _ | Closed ->
+            Lwt.wakeup_later_exn res_r Retry;
+            Lwt.return_unit
+        | Failed e ->
+            Lwt.wakeup_later_exn res_r e;
+            Lwt.return_unit
+        | Connecting _ -> assert false
+        end >>= fun () ->
+      res
+
+    let create
+      ?(finalise=fun _ -> Lwt.return_unit)
+      ?(ctx=Net.default_ctx)
+      uri
+      =
+      let mutex = Lwt_mutex.create () in
+      Lwt.async (fun () -> Lwt_mutex.lock mutex);
+      let connection =
+        Net.connect_uri ~ctx uri >|= fun (_, ic, oc) ->
+        (ic, oc)
+      in
+      let pipeline =
+        { finalise
+        ; in_flight = Queue.create ()
+        ; conn = Connecting connection
+        ; mutex
+        ; condition = Lwt_condition.create ()
+        }
+      in
+      Lwt.on_any connection
+        begin fun connection ->
+          pipeline.conn <- Full connection;
+          Lwt_mutex.unlock pipeline.mutex;
+          Lwt.async (fun () -> reader pipeline);
+        end
+        begin fun e ->
+          pipeline.conn <- Failed e;
+          Lwt_mutex.unlock pipeline.mutex;
+        end;
+      pipeline
+  end
+
+  module Conn_cache =
+  struct
+    type t = < process : ?body:Body.t -> Request.t -> (Response.t * Body.t) Lwt.t >
+    module Uri_map = Map.Make (Uri)
+
+    let no_cache ?ctx:ctx () :t = object
+      method process ?body req =
+        let pipeline = Pipeline.create ?ctx (Request.uri req) in
+        let res = Pipeline.process pipeline ?body req in
+        (* this can be simplified when https://github.com/mirage/ocaml-conduit/pull/319 is released. *)
+        Lwt.async begin fun () ->
+          res >>= fun (_, body) ->
+          match body with
+          | `Empty | `String _ | `Strings _ -> Lwt.return_unit
+          | `Stream stream ->
+              Lwt_stream.closed stream >>= fun () ->
+              Pipeline.shutdown pipeline
+          end;
+        res
+    end
+
+    (*
+     * Select from existing connections by hostname, not IP.
+     * One hostname may resolve to multiple IP addresses.
+     * In such a case we want only one connection.
+     * The downside is that some resolver redirections may break.
```

Comment:
Indeed. If you don't use the connection cache this doesn't matter. But as I wrote initially, I suspect using a `Conduit.endp` instead of an hostname for opening connections is better.
At the moment I do not think there is much to gain from making it configurable. If you have special needs in that you want multiple connections to the very same IP you are probably better of directly using the pipeline.

Reply:
I don’t have a strong opinion right now, let’s see if somebody else has something to say about it

Reply:
In my latest changes I indeed changed the type of the connection cache keys to `endp`. This indeed only makes a difference in case you are using a connection cache.